docs/architecture/adr-009-transport-agnostic-peer-abstraction.md

```diff
@@ -0,0 +1,194 @@
+spin up and plug into the machinery which lifts bytes from the wire into the
+core domain and transports messages into the rest of the system.
+
+## Decision
+
+### Transport
+
+Wraping the design is the `Transport`. Modelled with the properties of
+a physical network endpoint in mind, which can be bound and stopped. It should
+strongly correspond to the acquisition and lifecycle management of network
+resources on the system.
+
+``` rust
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;
+
+    fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;
+    fn shutdown(&self) -> Result<()>;
+}
+```
+
+After the successful bind the caller holds an `Endpoint` as well as a stream of
+incoming `Connection`s. Which is a standardised way to connect to new peers and
+react to newly connected ones respectively.
+
+``` rust
+pub trait Endpoint: Send {
+    type Connection;
+
+    fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;
+    fn listen_addrs(&self) -> SocketAddr;
+}
+```
+
+Centerpiece of the whole shebang is the `Connection`. It represents a connected
+peer and provides the primitives to get data and send data from a peer. It is
+designed with the outlook to support stream based trasnports down the road.
+While being open to enable feature parity with current production installations
+based on tendermint-go's `MConn`.
+
+``` rust
+pub trait Connection: Send {
+    type Error: 'static + fmt::Display + std::error::Error + Send + Sync;
+    type Read: Read;
+    type Write: Write;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr>;
+    fn close(&self) -> Result<()>;
+    fn local_addr(&self) -> SocketAddr;
+    fn open_bidirectional(
+        &self,
+        stream_id: &StreamId,
+    ) -> Result<(Self::Read, Self::Write), Self::Error>;
+    fn public_key(&self) -> PublicKey;
+    fn remote_addr(&self) -> SocketAddr;
+}
+```
+
+### Peer
+
+Given a correct implementation of a `Transport` and its `Connection` newly
+established ones will be wrapped with a `Peer`. Which is in charge of setting
+up the correct streams on the `Connection` and multiplex messages - incoming
+and outgoing alike - efficiently. It's also an attempt to enforce
+correct-by-construction constraints on the state machine of the peer. To avoid
+misuse or unexpected transitions. The only way to construct is, is from an
+existing connection which gives the caller a connected peer. When invoking run
+on that one a fully function peer is "returned". Therefore the states look
+like: `Connected -> Running -> Stopped`.
+
+``` rust
+impl<Conn> Peer<Connected<Conn>>
+where
+    Conn: Connection,
+{
+    pub fn run(self, stream_ids: Vec<StreamId>) -> Result<Peer<Running<Conn>>> {
+        // ...
+    }
+
+    fn stop(self) -> Result<Peer<Stopped>> {
```

Comment:
why it's possible to stop `Connected` (not `Running`) peer?

Reply:
Both `Connected` and `Running` have an associated `stop` method. :)

Reply:
I forgot to include it in the ADR, the code has it.

Reply:
Actually it's in there. See above.

Reply:
> Both Connected and Running have an associated stop method. :)

yeah, so my question is why it's possible to stop a non-running peer? it seems weird to me

Reply:
Ah right, I miss-read your question. From my perspective a `Connected` peer already holds onto some resources (i.e. the connection)  and maybe there is a valid use-case where it shouldn't be transitioned into running state but dropped before.
While we can rely on the `Drop` semantics alone I wanted a transition in the terminal `Stopped` state for both "starting" states.